Brokerages/Binance/BinanceBrokerage.Streaming.cs

```diff
@@ -0,0 +1,86 @@
+﻿using Newtonsoft.Json.Linq;
+using QuantConnect.Logging;
+using RestSharp;
+using System;
+using System.IO;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace QuantConnect.Brokerages.Binance
+{
+    /// <summary>
+    /// UserData Session class used to manage listen key
+    /// </summary>
+    public partial class BinanceBrokerage
+    {
+        private object _listenKeyLocker = new object();
+        private string userDataStreamEndpoint = "/api/v1/userDataStream";
+
+        /// <summary>
+        /// Represents UserData Session listen key
+        /// </summary>
+        public string SessionId { get; private set; }
+
+        /// <summary>
+        /// Check User Data stream listen key is alive
+        /// </summary>
+        /// <returns></returns>
+        private bool SessionKeepAlive()
+        {
+            if (string.IsNullOrEmpty(SessionId))
+            {
+                throw new Exception("BinanceBrokerage:UserStream. listenKey wasn't allocated or has been refused.");
+            }
+
+            var ping = new RestRequest(userDataStreamEndpoint, Method.PUT);
+            ping.AddHeader(KeyHeader, ApiKey);
+            ping.AddParameter(
+                "application/x-www-form-urlencoded",
+                Encoding.UTF8.GetBytes($"listenKey={SessionId}"),
+                ParameterType.RequestBody
+            );
+
+            var pong = ExecuteRestRequest(ping);
+            return pong.StatusCode == HttpStatusCode.OK;
+        }
+
+        /// <summary>
+        /// Stops the session
+        /// </summary>
+        public void StopSession()
+        {
+            if (!string.IsNullOrEmpty(SessionId))
```

Comment:
Is the inverse an exception? In other words, if `string.IsNullOrEmpty(SessionId)` -- is that an error case such as treated in `SessionKeepAlive()` ?

Reply:
hmm.. should we throw exception if SessionId is null or empty? In my mind, there is no any point to do that as it's already empty (matches to our expectations), that's why silence in this case.

Reply:
I would say yes since it implies the consumer of this class _thinks_ there's an open session and is trying to stop it. It's like trying to dispose an object that was never initialized. This is not a critical issue though, more of a style/preference around temporal coupling between methods in a class.